src/github.com/couchbaselabs/sync_gateway/base/bucket.go

```diff
@@ -105,6 +130,123 @@ func (bucket couchbaseBucket) StartTapFeed(args walrus.TapArguments) (walrus.Tap
 	return &impl, nil
 }
 
+// Start cbdatasource-based DCP feed.  Supports
+func (bucket couchbaseBucket) StartDCPFeed(args walrus.TapArguments) (walrus.TapFeed, error) {
+
+	// Recommended usage of cbdatasource is to let it manage it's own dedicated connection, so we're not
+	// reusing the bucket connection we've already established.
+	urls := []string{bucket.spec.Server}
+	poolName := bucket.spec.PoolName
+	if poolName == "" {
+		poolName = "default"
+	}
+	bucketName := bucket.spec.BucketName
+
+	vbucketIdsArr := []uint16(nil) // nil means get all the vbuckets.
+
+	receiver := NewDCPReceiver()
+
+	maxVbno, err := bucket.getMaxVbno()
+	if err != nil {
+		return nil, err
+	}
+
+	startSeqnos := make(map[uint16]uint64, maxVbno)
+	vbuuids := make(map[uint16]uint64, maxVbno)
+
+	// GetStatsVbSeqno retrieves high sequence number for each vbucket, to enable starting
+	// DCP stream from that position.  Also being used as a check on whether the server supports
+	// DCP.
+	statsUuids, highSeqnos, err := bucket.GetStatsVbSeqno(maxVbno)
+	if err != nil {
+		return nil, errors.New("Error retrieving stats-vbseqno - DCP not supported")
+	}
+
+	if args.Backfill == walrus.TapNoBackfill {
+		// For non-backfill, use vbucket uuids, high sequence numbers
+		LogTo("Feed+", "Seeding seqnos: %v", highSeqnos)
+		vbuuids = statsUuids
+		startSeqnos = highSeqnos
+	}
+	receiver.SeedSeqnos(vbuuids, startSeqnos)
+
+	// Auth handler for DCP connection needs to support bucket name defaults
+	username, password := "", ""
+	if bucket.spec.Auth != nil {
+		username, password, _ = bucket.spec.Auth.GetCredentials()
+	}
+	auth := &dcpAuth{username, password, bucketName}
+
+	LogTo("Feed+", "Connecting to new bucket datasource.  URLs:%s, pool:%s, name:%s, auth:%s", urls, poolName, bucketName, auth)
+	bds, err := cbdatasource.NewBucketDataSource(urls, poolName, bucketName, "", vbucketIdsArr, auth, receiver, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	events := make(chan walrus.TapEvent)
+	impl := couchbaseDCPFeedImpl{bds, events}
+
+	if err = bds.Start(); err != nil {
+		return nil, err
+	}
+
+	go func() {
+		for dcpEvent := range receiver.EventFeed {
+			events <- dcpEvent
+		}
+	}()
+
+	return &impl, nil
+}
+
+func (bucket couchbaseBucket) GetStatsVbSeqno(maxVbno uint16) (map[uint16]uint64, map[uint16]uint64, error) {
```

Comment:
I think we should figure out some way to prevent having these two identical anonymous map return values.   Using named return values here would at least document what the return values contain.  (and although named return values have their own set of issues, I think it would be the lesser of two evils)